File: src/api/video.ts
/**
 * 后端命令封装（与 src-tauri/src/commands/video.rs 对齐）
 *
 * 前端唯一允许直接调用 invoke 的地方；字段统一 camelCase，
 * 与 Rust 侧 `#[serde(rename_all = "camelCase")]` 一一对应。
 */
import { invoke } from '@tauri-apps/api/core'
import { listen, type UnlistenFn } from '@tauri-apps/api/event'

/** 解析出的视频元信息 */
export interface VideoInfo {
  awemeId: string
  title: string
  desc: string
  author: string
  durationMs: number
  cover: string
  playUrl: string
}

/** 下载进度事件负载 */
export interface DownloadProgress {
  taskId: string
  downloaded: number
  total: number
}

/** 下载完成事件负载 */
export interface DownloadDone {
  taskId: string
  path: string
}

/** 下载失败事件负载 */
export interface DownloadError {
  taskId: string
  error: string
}

/** 主页作品列表中的一条 */
export interface PostItem {
  awemeId: string
  desc: string
  author: string
  durationMs: number
  cover: string
  createTime: number
  diggCount: number
  commentCount: number
  shareCount: number
  collectCount: number
  playCount: number
}

/** 主页作品列表（分页） */
export interface PostListResult {
  items: PostItem[]
  hasMore: boolean
  maxCursor: number
}

/** 启动下载所需参数；playUrl 为空时后端按 awemeId 自动解析（批量任务） */
export interface StartDownloadParams {
  taskId: string
  playUrl?: string
  title: string
  awemeId: string
  platform: string
  quality: string
}

/** 解析分享文本/链接 */
export function parseVideo(text: string): Promise<VideoInfo> {
  return invoke<VideoInfo>('parse_video', { text })
}

/** 启动后台下载任务（立即返回，进度走事件） */
export function startDownload(p: StartDownloadParams): Promise<void> {
  return invoke('start_download', {
    taskId: p.taskId,
    playUrl: p.playUrl ?? '',
    title: p.title,
    awemeId: p.awemeId,
    platform: p.platform,
    quality: p.quality,
  })
}

/** 拉取用户主页作品列表（a_bogus 由前端对同一组 query 签名） */
export function fetchUserPosts(
  secUserId: string,
  aBogus: string,
  maxCursor?: number | null,
): Promise<PostListResult> {
  return invoke<PostListResult>('fetch_user_posts', {
    secUserId,
    aBogus,
    maxCursor: maxCursor ?? null,
  })
}

/** 暂停下载任务（中断网络请求，保留 .part 临时文件） */
export function cancelDownload(taskId: string): Promise<void> {
  return invoke('cancel_download', { taskId })
}

/** 订阅下载进度事件，返回取消监听函数 */
export function onDownloadProgress(cb: (p: DownloadProgress) => void): Promise<UnlistenFn> {
  return listen<DownloadProgress>('download-progress', (e) => cb(e.payload))
}

/** 订阅下载完成事件 */
export function onDownloadDone(cb: (p: DownloadDone) => void): Promise<UnlistenFn> {
  return listen<DownloadDone>('download-done', (e) => cb(e.payload))
}

/** 订阅下载失败事件 */
export function onDownloadError(cb: (p: DownloadError) => void): Promise<UnlistenFn> {
  return listen<DownloadError>('download-error', (e) => cb(e.payload))
}

File: src-tauri/src/commands/video.rs
//! 视频解析 / 下载命令
//!
//! 所有命令都从设置文件读取配置（保存目录、文件名规则、代理、断点续传），
//! 保证「设置页改了什么，下载就按什么来」。
//!
//! 下载为后台任务：`start_download` 立即返回，进度/完成/失败通过事件推送前端：
//! - `download-progress` { taskId, downloaded, total }
//! - `download-done`     { taskId, path }
//! - `download-error`    { taskId, error }

use std::collections::HashMap;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Mutex};

use tauri::{AppHandle, Emitter, Manager};

use crate::config::{load_config, AppSettings};
use crate::download::{self, FilenameRule};
use crate::http::{self, ProxyCfg};
use crate::platforms::{self, PostListResult, VideoInfo};

/// 共享的下载状态：task_id → 中止标志，用于暂停
#[derive(Default)]
pub struct DownloadState {
    aborts: Mutex<HashMap<String, Arc<AtomicBool>>>,
}

/// 读取设置，并在保存目录为空时回填系统下载目录
fn resolve_settings(app: &AppHandle) -> AppSettings {
    let mut settings = app
        .path()
        .app_config_dir()
        .map(|d| d.join("settings.json"))
        .map(|p| load_config(&p))
        .unwrap_or_default();
    if settings.save_dir.is_empty() {
        if let Ok(d) = app.path().download_dir() {
            settings.save_dir = d.to_string_lossy().to_string();
        }
    }
    settings
}

/// 从配置构造代理设置
fn proxy_from(settings: &AppSettings) -> ProxyCfg {
    ProxyCfg {
        enabled: settings.proxy_enabled,
        host: settings.proxy_host.clone(),
        port: settings.proxy_port.clone(),
    }
}

/// 解析抖音分享文本/链接，返回视频元信息
#[tauri::command]
pub async fn parse_video(app: AppHandle, text: String) -> Result<VideoInfo, String> {
    let settings = resolve_settings(&app);
    let proxy = proxy_from(&settings);
    platforms::parse(&text, &proxy).await
}

/// 启动一个下载任务（后台异步执行，立即返回）
///
/// `play_url` 为空时视为「批量任务」（前端 enqueueBatch 创建、没有解析出的播放地址）：
/// 后台先按 aweme_id 走 `platforms::parse` 拿到无水印地址，再进入统一下载流程。
#[tauri::command]
pub fn start_download(
    app: AppHandle,
    task_id: String,
    play_url: String,
    title: String,
    aweme_id: String,
    platform: String,
    quality: String,
) -> Result<(), String> {
    // 解析配置
    let settings = resolve_settings(&app);
    let proxy = proxy_from(&settings);
    let client = http::http_client(&proxy)?;
    let save_dir = settings.save_dir;
    let rule = FilenameRule::from_str(&settings.filename_rule);
    let resume = settings.resume;
    let notify_done = settings.notify_done;

    // 注册中止标志，供暂停使用（解析也发生在后台，暂停在解析期间即可置位）
    let abort = Arc::new(AtomicBool::new(false));
    {
        let state = app.state::<DownloadState>();
        state
            .aborts
            .lock()
            .unwrap()
            .insert(task_id.clone(), abort.clone());
    }

    // 后台跑：先补全 play_url（批量任务），再下载。结束后清理中止标志。
    // 进度/完成/失败事件由 download::run 内部发出，暂停无需事件（前端已置 paused）。
    let app2 = app.clone();
    let tid = task_id.clone();
    let pid = aweme_id.clone();
    let proxy2 = proxy.clone();
    tauri::async_runtime::spawn(async move {
        // 批量任务没有播放地址 → 先解析出无水印地址
        let play_url = if play_url.is_empty() {
            match platforms::parse(&format!("https://www.douyin.com/video/{pid}"), &proxy2).await {
                Ok(info) => info.play_url,
                Err(e) => {
                    let _ = app2.emit(
                        "download-error",
                        download::DownloadError {
                            task_id: tid.clone(),
                            error: format!("解析视频地址失败: {e}"),
                        },
                    );
                    let state = app2.state::<DownloadState>();
                    state.aborts.lock().unwrap().remove(&tid);
                    return;
                }
            }
        } else {
            play_url
        };

        download::run(
            &app2,
            &client,
            &tid,
            &play_url,
            &title,
            &aweme_id,
            &save_dir,
            rule,
            &platform,
            &quality,
            resume,
            notify_done,
            abort,
        )
        .await;

        let state = app2.state::<DownloadState>();
        state.aborts.lock().unwrap().remove(&tid);
    });

    Ok(())
}

/// 拉取用户主页作品列表（分页，20 条/页）。
/// `a_bogus` 由前端用 abogus.js 对同一组 query 参数签名后传入。
#[tauri::command]
pub async fn fetch_user_posts(
    app: AppHandle,
    sec_user_id: String,
    a_bogus: String,
    max_cursor: Option<u64>,
) -> Result<PostListResult, String> {
    let settings = resolve_settings(&app);
    let proxy = proxy_from(&settings);
    let client = http::http_client(&proxy)?;
    platforms::douyin::fetch_user_posts(&client, &proxy, &sec_user_id, &a_bogus, max_cursor).await
}

/// 暂停任务：置位中止标志，后台下载循环检测到后中断并保留 `.part`
#[tauri::command]
pub fn cancel_download(app: AppHandle, task_id: String) -> Result<(), String> {
    let state = app.state::<DownloadState>();
    if let Some(flag) = state.aborts.lock().unwrap().get(&task_id) {
        flag.store(true, Ordering::Relaxed);
    }
    Ok(())
}

/// 获取系统默认下载目录（首次打开设置时兜底用）
#[tauri::command]
pub fn get_default_dir(app: AppHandle) -> Result<String, String> {
    let path = app
        .path()
        .download_dir()
        .map_err(|e| format!("获取下载目录失败: {e}"))?;
    Ok(path.to_string_lossy().to_string())
}
